### PR TITLE
perf(shwap/bitswap): add session pools inside Getter

### PR DESCRIPTION
**Pull request by Wondertan**
In the Robusta environment this change:
* Improved load distribution across BN/FNs
* Sped up sampling

Originally, I was testing Bitswap with this change but decided against including it in the patch, as it is solely an optimization. However, without this change on the rc tag, load distribution was rather strange. One LN sampled at 180/s, while all 4 others at 4/s. Adding pooling fixed the issue and resulted in all 5 LNs at about 100/s.

### CODE REVIEW

File: share/shwap/p2p/bitswap/getter.go [hunk 1]
```diff
@@ -57,9 +58,21 @@ func NewGetter(
...
 func (g *Getter) Start() {
 	ctx, cancel := context.WithCancel(context.Background())
-	g.availableSession = g.exchange.NewSession(ctx)
-	g.archivalSession = g.exchange.NewSession(ctx)
 	g.cancel = cancel
+
+	var availableID atomic.Uint64
+	g.availablePool.New = func() interface{} {
+		log.Debugw("creating new available window session", "id", availableID.Load())
+		defer availableID.Add(1)
```
Comment (cristaloleg) [inline share/shwap/p2p/bitswap/getter.go:66]: Why `defer` here?

File: share/shwap/p2p/bitswap/getter.go [hunk 2]
```diff
@@ -230,17 +258,37 @@ func (g *Getter) GetNamespaceData(
...
+	v := g.availablePool.Get()
+	if v == nil {
+		panic("Getter must be started")
+	}
+
+	return v.(exchange.Fetcher)
+}
+
+// poolSession puts session back into the respective session pool
+func (g *Getter) poolSession(session exchange.Fetcher, isArchival bool) {
```
Comment (cristaloleg) [inline share/shwap/p2p/bitswap/getter.go:286]: Nit: more common naming will be `putSession`.

File: share/shwap/p2p/bitswap/getter.go [hunk 3]
```diff
@@ -230,17 +258,37 @@ func (g *Getter) GetNamespaceData(
...
+// isArchival reports whether the header is for archival data
+func (g *Getter) isArchival(hdr *header.ExtendedHeader) bool {
+	return !availability.IsWithinWindow(hdr.Time(), g.availWndw)
+}
 
-	isWithinAvailability := availability.IsWithinWindow(hdr.Time(), g.availWndw)
-	if isWithinAvailability {
-		session = g.availableSession
+// session takes a session out of the respective session pool
+func (g *Getter) session(isArchival bool) exchange.Fetcher {
```
Comment (cristaloleg) [inline share/shwap/p2p/bitswap/getter.go:267]: Nit: same-ish here, why not `get-/acquire-`?

File: share/shwap/p2p/bitswap/getter.go (hunk 2 above)
Comment (Wondertan) [reply share/shwap/p2p/bitswap/getter.go:286]: Sure

File: share/shwap/p2p/bitswap/getter.go (hunk 1 above)
Comment (cristaloleg) [reply share/shwap/p2p/bitswap/getter.go:66]: So up to you.

Comment (Wondertan) [reply share/shwap/p2p/bitswap/getter.go]: Oh, you're right. Removed defer

File: share/shwap/p2p/bitswap/getter.go [hunk 4]
```diff
@@ -230,17 +258,37 @@ func (g *Getter) GetNamespaceData(
 	return nsShrs, nil
 }
 
-// session decides which fetching session to use for the given header.
-func (g *Getter) session(ctx context.Context, hdr *header.ExtendedHeader) exchange.Fetcher {
-	session := g.archivalSession
+// isArchival reports whether the header is for archival data
+func (g *Getter) isArchival(hdr *header.ExtendedHeader) bool {
+	return !availability.IsWithinWindow(hdr.Time(), g.availWndw)
+}
```
Comment (Wondertan) [inline share/shwap/p2p/bitswap/getter.go:264]: btw, isArchival is much more readable, shorter and easier to work with then IsWithinWindow

**Review by cristaloleg — APPROVED** [review]
SGTM

File: share/shwap/p2p/bitswap/getter.go [hunk 5]
```diff
@@ -57,12 +58,24 @@ func NewGetter(
...
 func (g *Getter) Start() {
 	ctx, cancel := context.WithCancel(context.Background())
-	g.availableSession = g.exchange.NewSession(ctx)
-	g.archivalSession = g.exchange.NewSession(ctx)
 	g.cancel = cancel
+
+	var availableID atomic.Uint64
+	g.availablePool.New = func() interface{} {
+		log.Debugw("creating new available window session", "id", availableID.Load())
+		availableID.Add(1)
```
Comment (walldiss) [inline share/shwap/p2p/bitswap/getter.go:66]: Loaded value might be overwriten between add, so logging might be confusing. Would suggest rather to call  atomic.Add() first and use value returned by Add() for logging. 
```suggestion
	g.availablePool.New = func() interface{} {
	        id := availableID.Add(1)
		log.Debugw("creating new available window session", "id", id)
		
```

**Comment by Wondertan** [comment]
I didn't realize that `sync.Pool` can silently delete objects, so there is a bug - once the pool deletes a session, it's not cleaned up from the bitswap itself. Also, we don't need to create many more sessions but require a constant amount equal to a number of DASer workers, so instead, we should implement pooling over the map.

@walldiss is gonna take over that.

**Review by Wondertan — COMMENTED** [review]
Thanks @walldiss. LGTM

**Review by walldiss — DISMISSED** [review]
approving on behalf of @Wondertan

**Review by walldiss — APPROVED** [review]

**Review by vgonkivs — APPROVED** [review]

**Pull request merged**